### PR TITLE
Lglen/sha1solver

### PR DESCRIPTION
**Pull request by lukasz-glen**

### CODE REVIEW

File: sha1solver/sha1solver.c [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+{
+  int r = (int) strtol(argv[1], (char **)NULL, 10);
+  int s = (int) strtol(argv[2], (char **)NULL, 10);
+
+  int region_start = (int) strtol(argv[3], (char **)NULL, 10);
+  int region_end = (int) strtol(argv[4], (char **)NULL, 10);
+
+  char *pattern = argv[5];
+
+  FILE *f_in = fopen("infile", "rb");
```
Comment (mplebanski) [inline sha1solver/sha1solver.c]: Why input file is fixed? It could be one of the input parameters

File: sha1solver/README.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+### Input file
+
+The file `infile` is an input binary file. The name of the file cannot be changed. 
+
+### Execution and output
+
+You should run multiple instances of the program. Each takes the same input file but different parameters. Each instance writes an answer to the output file `out.txt`, so watch out to not override it. The output file contains text. Each rows contains bytes from the region and the hash of the `infile`.
+
+The expected output (content of the `out.txt` file) contains rows like
```
Comment (mplebanski) [inline sha1solver/README.md]: What is the set of parameters that produces this output?

Comment (mplebanski) [reply sha1solver/README.md]: I'm trying to achieve the same results.

**Comment by mplebanski** [comment]
@lukasz-glen Please run it through some styles of clang-format because for example indents are not consistent.

File: sha1solver/sha1solver.c [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    SHA1_CTX * ctx,
+    char *pattern,
+    int pattern_len,
+    int r,
+    int s,
+    int first)
+{
+    unsigned int ii;
+
+    SHA1Init(ctx);
```
Comment (mplebanski) [inline sha1solver/sha1solver.c]: Consider moving this initilization out of this function

Comment (mplebanski) [reply sha1solver/sha1solver.c]: Partially because there is such initilization happening in SHA1Check already.

File: sha1solver/sha1solver.c [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+int SHA1Progress(
+    char *hash_out,
+    char *str,
+    int len,
+    int poz_start,
+    int poz_end,
+    SHA1_CTX * ctx,
+    char *pattern,
+    int pattern_len,
+    int r,
```
Comment (mplebanski) [inline sha1solver/sha1solver.c]: What is `r` and `s`? Can more human-readable names be applied?

File: sha1solver/sha1solver.c [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    FILE * f_out)
+{
+    SHA1_CTX ctx;
+    unsigned int ii;
+    char result[21];
+    char hexresult[41];
+    hexresult[40] = '\0';
+    size_t offset;
+    int char_pattern_len = pattern_len/2;
+    char *char_pattern = (char*)malloc ( (char_pattern_len) * sizeof (char));
```
Comment (mplebanski) [inline sha1solver/sha1solver.c]: [There is no need to cast result of malloc](https://stackoverflow.com/a/605858/2429333). Also isn't type `char *` for parameter `pattern` implying that `pattern_len` will be of appropriate length? That could simplify this expression to:
```c
char *char_pattern = malloc(char_pattern_len);
```

File: sha1solver/sha1solver.c [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    int poz_end,
+    char *pattern,
+    int pattern_len,
+    int r,
+    int s,
+    FILE * f_out)
+{
+    SHA1_CTX ctx;
+    unsigned int ii;
+    char result[21];
```
Comment (mplebanski) [inline sha1solver/sha1solver.c]: Consider moving `21` magic number to a named macro definition.

Comment (mplebanski) [reply sha1solver/sha1solver.c]: For readability ofc.

File: sha1solver/sha1solver.c [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    int char_pattern_len = pattern_len/2;
+    char *char_pattern = (char*)malloc ( (char_pattern_len) * sizeof (char));
+
+    for (int ii = 0; ii < char_pattern_len; ii += 1) {
+	char_pattern[ii] = hex(pattern+2*ii)*16+hex(pattern+2*ii+1);
+    }
+
+    result[20] = '\0';
+    for (ii=poz_start; ii<poz_end; ii+=1)
+        str[ii] = (const unsigned char) 0;
```
Comment (mplebanski) [inline sha1solver/sha1solver.c]: A shorter notation is:
```
str[ii] = '\0';
```

File: sha1solver/sha1solver.c [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+    int poz_start,
+    int poz_end,
+    SHA1_CTX * ctx,
+    char *pattern,
+    int pattern_len,
+    int r,
+    int s,
+    int first)
+{
+    unsigned int ii;
```
Comment (mplebanski) [inline sha1solver/sha1solver.c]: Consider moving this declaration to loop initialization statement. That will make clear for each loop this variable is used. Note that you are doing it somewhere anyway.

File: sha1solver/sha1solver.c [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+
+    SHA1Init(ctx);
+    for (ii=0; ii<poz_start; ii+=1)
+        SHA1Update(ctx, (const unsigned char*)str + ii, 1);
+
+    SHA1_CTX ctx1;
+    int poz;
+    poz = poz_start;
+    if (!first) {
```
Comment (mplebanski) [inline sha1solver/sha1solver.c:37]: What this part of the code is doing?

File: sha1solver/sha1solver.c [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+	str[poz] ++;
+        if (str[poz] == (const unsigned char) 0) {
+            poz ++;
+        }
+    }
+    while (1) {
```
Comment (mplebanski) [inline sha1solver/sha1solver.c:43]: In simple words, when does the loop terminate?

File: sha1solver/sha1solver.c [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+    result[20] = '\0';
+    for (ii=poz_start; ii<poz_end; ii+=1)
+        str[ii] = (const unsigned char) 0;
+
+    SHA1Init(&ctx);
+    for (ii=0; ii<poz_start; ii+=1)
+        SHA1Update(&ctx, (const unsigned char*)str + ii, 1);
+
+    int first = 1;
+    while (SHA1Progress(result, str, len, poz_start, poz_end, &ctx, char_pattern, char_pattern_len, r, s, first) == 0) {
```
Comment (mplebanski) [inline sha1solver/sha1solver.c]: From what I see the `first` parameter could be removed if `str` and `poz_start` were configured at this level.

File: sha1solver/sha1solver.c (hunk 3 above)
Comment (lukasz-glen) [reply sha1solver/sha1solver.c]: Yep, there was a C&P error.

File: sha1solver/README.md (hunk 2 above)
Comment (lukasz-glen) [reply sha1solver/README.md]: corrected

File: sha1solver/sha1solver.c (hunk 6 above)
Comment (lukasz-glen) [reply sha1solver/sha1solver.c]: done

File: sha1solver/sha1solver.c (hunk 7 above)
Comment (lukasz-glen) [reply sha1solver/sha1solver.c]: works, done

File: sha1solver/sha1solver.c (hunk 8 above)
Comment (lukasz-glen) [reply sha1solver/sha1solver.c]: done, but that is obvious for one of the loops.

File: sha1solver/sha1solver.c (hunk 11 above)
Comment (lukasz-glen) [reply sha1solver/sha1solver.c]: actually poz_start is constant - it is starting byte position of the region

Comment (lukasz-glen) [reply sha1solver/sha1solver.c]: I am not sure what you mean

File: sha1solver/sha1solver.c (hunk 5 above)
Comment (lukasz-glen) [reply sha1solver/sha1solver.c]: done

File: sha1solver/sha1solver.c (hunk 1 above)
Comment (lukasz-glen) [reply sha1solver/sha1solver.c]: It is parameter now along with optional output file name

File: sha1solver/sha1solver.c (hunk 9 above)
Comment (lukasz-glen) [reply sha1solver/sha1solver.c:37]: I have added comments in the code

File: sha1solver/sha1solver.c (hunk 10 above)
Comment (lukasz-glen) [reply sha1solver/sha1solver.c:43]: I have added comments in the code

File: sha1solver/sha1solver.c (hunk 4 above)
Comment (lukasz-glen) [reply sha1solver/sha1solver.c]: I have added more comments

File: sha1solver/sha1solver.c [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+ * Searches for the next configuration of bytes in the region such that hash of str matches the given pattern.
+ * It starts with current configuration of bytes in the region, increments the bytes until first match is found.
+ * It modifies the region in str, actually the region in str is kind of state of the search for all matches.
+ * The function is to be called multiple times.
+ * The bytes in the region is interpreted as little endian representation of the number. 
+ * Upon the first call it should be 0. The search is done when the number loops back to 0.
+ * Not all configuration of bytes are eligible. Only one s-th part is eligible. Which part is set by param r.
+ * @param poz_start first position of the region (inclusive)
+ * @param poz_end last position of the region (exclusive)
+ * @param r reminder
```
Comment (mplebanski) [inline sha1solver/sha1solver.c]: Actually it's `remainder`*

**Review by mplebanski — APPROVED** [review]

**Pull request merged**